### ISSUE TITLE
How to send transport parameters in runtime with AcMailer 7?

### ISSUE THREAD

**Issue by Syberon** [acelaya/ZF-AcMailer#222]
Hello.

Before now i used a AcMailer 6 in my project and sets transport parameters in runtime via event listener by this function:

`
    public function onPreSend(MailEvent $e)
    {
        if (!$this->mailService) {
            $this->mailService = $e->getMailService();
            $params = $this->modelParams->verticalSlice('value', 'id');

            if ($params['mail_transport'] == "Smtp") {
                $options = [
                    'connection_class'  => 'login',
                    'host'              => $params['mail_server'],
                    'port'              => $params['mail_port'] ? $params['mail_port'] : 25,
                    'connection_config' => [
                        'username' => $params['mail_username'],
                        'password' => $params['mail_password'],
                        'ssl'      => $params['mail_ssl'] ? 'ssl' : null
                    ]
                ];
                $this->mailService->setTransport(new Smtp(new SmtpOptions($options)));
            } else {
                $this->mailService->setTransport(new Sendmail());
            }
        }
        return true;
    }
`

But in process of migration to version 7 i found that now i can't get/set transport via mailService methods. How can i realize with version 7?

**Comment by acelaya** [comment]
I consider setters on services a bad design decision, that's why I removed all of them from the service for v7.

You should probably consider wrapping that logic somehow, but not in a mail event listener, since that's not what they are meant for.

I see a couple options, but I would need a little bit more information on what you want to achieve.
* Is it only two transport configs you want to use?
* Is this `modelParams` object fetching the config dynamically from somewhere?

**Comment by Syberon** [comment]
Thanks for fast response.

At first, this answers to your questions:
1. Yes, i use only Sendmail and Smtp transport.
2. `modelParams` is object that fetch the parameters from database.

Today i tried to resolve my problem with using of custom transport factory instead of event listener. I created a custom transport factory and moved all this logic to it. This factory returns Sendmail() or Smtp() transport object initialized with my dynamic parameters.

In services config:
`'mailer.custom.transport' => MailTransportFactory.php:`

In factory file:
```
$modelParams = $container->get(System::class);
$params = $modelParams->verticalSlice('value', 'id');
if ($params['mail_transport'] == "Smtp") {
    $options = [
        'connection_class'  => 'login',
            'host'              => $params['mail_server'],
            'port'              => $params['mail_port'] ? $params['mail_port'] : 25,
            'connection_config' => [
                'username' => $params['mail_username'],
                'password' => $params['mail_password'],
                'ssl'      => $params['mail_ssl'] ? 'ssl' : null
            ]
    ];
    return new Smtp(new SmtpOptions($options));
} else {
    return new Sendmail();
}
```
Next i changed mailer config to use my custom service '**mailer.custom.transport**' as default mail transport:
```
*
* Default value is 'sendmail'
*/
'transport' => 'mailer.custom.transport',
```

And now it's normally worked with Sendmail chosen in application settings, because it doesn't need any parameters, but if i need to use Smtp there are another problem - my transport parameters (that i setted up in my custom transport factory) is replaced by empty values (from default config files) in function `setupTransportConfig` that called during mail service initialization in factory `MailServiceAbstractFactopy.php`. This behavior appears because my custom factory returning the instance of standard Transport\Smtp object and mailer function doesn't know that it created and initialized by my custom factory and tries to init this instance with default config parameters.

To prevent this behavior i think there is needed some option (flag) in config file that tells mail service factory about custom transport factory used. Maybe someting like '`custom_transport = true`' and prevents to initialize it with default config paramters.

**Comment by acelaya** [comment]
Wow! Amazing research :)

Using a custom transport service was one of the options I was thinking about.

The behavior you have found is totally a bug. When using a custom transport service, the factory should just fetch it as is. I think there's no need to add any extra config in order to detect if the service was defined, so I will try to provide a fix for this (I think I will have time tomorrow afternoon).

Thanks a lot for all the detailed information.

**Comment by acelaya** [comment]
@Syberon I have just released v7.3.3 which fixes the issue.

The fix was quite simple, you can see the pull request referencing this issue. Most of the new code is documentation and tests :)

**Comment by Syberon** [comment]
Yeah! Thanks)

Now i tested your solution and it works flawlessly.
You are doing a good work, Alejandro)
